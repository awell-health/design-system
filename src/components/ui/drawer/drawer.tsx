import { ReactNode, type FC } from 'react';
import clsx from 'clsx';

interface DrawerProps extends React.HTMLAttributes<HTMLDivElement> {
    isOpen: boolean
    children: ReactNode
    drawerWidth?: number
    side?: 'left' | 'right'
}


export const Drawer: FC<DrawerProps> = ({ isOpen, children, drawerWidth = 460, side = 'right', className }) => {

    return (
      <div className={
        clsx(
          'absolute top-0 h-full',
          side === 'left' ? 'left-0' : 'right-0',
          className
        )
      }
      style={{ width: `${drawerWidth}px` }}
      >
        <div className="absolute inset-0 overflow-hidden">
          <div
            className={clsx(
              'absolute h-full w-full top-0',
              side === 'left' ? 'left-0' : 'right-0',
              'bg-white',
              'transition-transform duration-300 ease-in-out',
              {
                'translate-x-0': isOpen && side === 'right',
                'translate-x-full': !isOpen && side === 'right',
                '-translate-x-0': isOpen && side === 'left',
                '-translate-x-full': !isOpen && side === 'left',
              }
            )}
          >
            <div className='w-full h-full relative'>
              <div className='h-full'>{children}</div>
            </div>
          </div>
        </div>
      </div>
    );
};
