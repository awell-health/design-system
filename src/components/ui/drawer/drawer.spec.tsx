import { expect, it, describe } from 'vitest';
import { render } from '@testing-library/react';
import { Drawer } from './drawer';
import React from 'react';

describe('Drawer', () => {
  const subject = (props = {}) => {
    const defaultProps = {
      isOpen: true,
      children: <div>Drawer Content</div>,
      drawerWidth: 460,
      side: 'right' as const
    };
    return render(<Drawer {...defaultProps} {...props} />);
  };

  it('matches snapshot', () => {
    expect(subject().container).toMatchSnapshot();
  });

  it('renders with correct width', () => {
    const { container } = subject({ drawerWidth: 500 });
    const drawer = container.firstChild as HTMLElement;
    expect(drawer).toHaveStyle({ width: '500px' });
  });

  it('positions on the right by default', () => {
    const { container } = subject();
    const drawer = container.firstChild as HTMLElement;
    expect(drawer).toHaveClass('right-0');
    expect(drawer).not.toHaveClass('left-0');
  });

  it('positions on the left when specified', () => {
    const { container } = subject({ side: 'left' });
    const drawer = container.firstChild as HTMLElement;
    expect(drawer).toHaveClass('left-0');
    expect(drawer).not.toHaveClass('right-0');
  });

  it('has correct classes when open', () => {
    const { container } = subject({ isOpen: true });
    const drawer = container.firstChild as HTMLElement;
    expect(drawer).toHaveClass('translate-x-0');
    expect(drawer).toHaveClass('shadow-lg');
    expect(drawer).not.toHaveClass('translate-x-full');
    expect(drawer).not.toHaveClass('shadow-none');
  });

  it('has correct classes when closed', () => {
    const { container } = subject({ isOpen: false });
    const drawer = container.firstChild as HTMLElement;
    expect(drawer).toHaveClass('translate-x-full');
    expect(drawer).toHaveClass('shadow-none');
    expect(drawer).not.toHaveClass('translate-x-0');
    expect(drawer).not.toHaveClass('shadow-lg');
  });

  it('has transition classes', () => {
    const { container } = subject();
    const drawer = container.firstChild as HTMLElement;
    expect(drawer).toHaveClass('transition-all');
    expect(drawer).toHaveClass('duration-300');
    expect(drawer).toHaveClass('ease-in-out');
  });
});