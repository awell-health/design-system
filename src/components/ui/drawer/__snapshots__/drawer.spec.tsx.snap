// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Drawer > matches snapshot 1`] = `
<div>
  <div
    class="absolute top-0 h-full bg-white transition-all duration-300 ease-in-out right-0 shadow-lg translate-x-0"
    style="width: 460px;"
  >
    <div>
      Drawer Content
    </div>
  </div>
</div>
`;
